app/src/main/java/org/mozilla/tv/firefox/pocket/PocketRepo.kt

```diff
@@ -0,0 +1,106 @@
+import android.os.SystemClock
+import android.support.annotation.UiThread
+import kotlinx.coroutines.experimental.CancellationException
+import kotlinx.coroutines.experimental.DefaultDispatcher
+import kotlinx.coroutines.experimental.Job
+import kotlinx.coroutines.experimental.delay
+import kotlinx.coroutines.experimental.launch
+import kotlinx.coroutines.experimental.withContext
+import org.mozilla.tv.firefox.BuildConfig
+import java.util.concurrent.TimeUnit
+
+private val CACHE_UPDATE_FREQUENCY_MILLIS = TimeUnit.MINUTES.toMillis(45)
+
+sealed class PocketRepoState {
+    data class LoadComplete(val videos: List<PocketFeedItem>) : PocketRepoState()
+    object Loading : PocketRepoState()
+    object NoKey : PocketRepoState()
+    object FetchFailed : PocketRepoState()
+}
+
+/**
+ * Manages backing state for Pocket data, as well as any logic related to
+ * retrieving or storing that data.
+ */
+open class PocketRepo(private val pocketEndpoint: PocketEndpoint) {
+
+    private val mutableState = MutableLiveData<PocketRepoState>()
+    open val state: LiveData<PocketRepoState> = mutableState
+
+    @Suppress("ANNOTATION_TARGETS_NON_EXISTENT_ACCESSOR") // Private properties generate fields so method annotations can't apply.
+    @get:UiThread
+    @set:UiThread
+    private var backgroundUpdates: Job? = null
+    @Volatile private var lastUpdateMillis = -1L
+    private val nextUpdateMillis get() = lastUpdateMillis + CACHE_UPDATE_FREQUENCY_MILLIS
+
+    init {
+        // mutableState.value should always be initialized at the top of init, because we treat
+        // it as non-null
+        @Suppress("SENSELESS_COMPARISON")
+        // Values of BuildConfig can change but the compiler doesn't know that
+        mutableState.value = when {
+            BuildConfig.POCKET_KEY == null -> PocketRepoState.NoKey
+            else -> PocketRepoState.Loading
+        }
+
+        update()
+    }
+
+    fun update() {
+        launch {
+            val fetchedState = requestVideos()
+            updateState(fetchedState)
+        }
+    }
+
+    @UiThread // update backgroundUpdates.
+    fun startBackgroundUpdates() {
+        backgroundUpdates?.cancel(CancellationException("Cancelling unexpectedly active job to ensure only one is running"))
+        backgroundUpdates = startBackgroundUpdatesInner()
+    }
+
+    @UiThread // update backgroundUpdates.
+    fun stopBackgroundUpdates() {
+        backgroundUpdates?.cancel(CancellationException("Stopping background updates"))
+        backgroundUpdates = null
+    }
+
+    private fun startBackgroundUpdatesInner() = launch {
+        while (true) {
+            val delayForMillis = nextUpdateMillis - SystemClock.elapsedRealtime()
+            if (delayForMillis > 0) {
+                delay(delayForMillis, TimeUnit.MILLISECONDS)
+            }
+
+            val newState = requestVideos()
+            // Delay next request if the previous was successful, otherwise try again immediately
```

Comment:
So we retry infinitely? That doesn't seem right. Maybe try to follow the previous logic, which I don't think was this?

Reply:
I agree that we should change this, but this is how it currently works in master.  I can make that change here or we can add a follow-up issue, either way is fine.

```kotlin
    private fun startBackgroundUpdatesInner() = launch {
        while (true) {
            val delayForMillis = nextUpdateMillis - SystemClock.elapsedRealtime()
            if (delayForMillis > 0) {
                delay(delayForMillis, TimeUnit.MILLISECONDS)
            }

            val deferredVideoUpdate = getRecommendedVideosNoCacheAsync()

            // We only want to update the cache 1) after the request completes so the user never has
            // to see a loading screen except for the initial load and 2) if the request has been
            // successful (i.e. non-null) so the user never has to see an error screen (unless we
            // never got data initially).
            val videoUpdate = deferredVideoUpdate.await()
            if (videoUpdate != null) {
                videosCache = deferredVideoUpdate
                lastUpdateMillis = SystemClock.elapsedRealtime()
            }
        }
    }
```

Reply:
So it is: I apologize for my assumption. 😞

I think one thing that makes this code unclear is that `nextUpdateMillis` is calculated from `lastUpdateMillis` so this infinite retry happens at most at intervals of 45 minutes, which seems fine to me (I had assumed it there was no delay because `nextUpdateMillis` wasn't updated). Since it's only used in one place, maybe we should calculate `nextUpdateMillis` at the top of the loop instead?

Reply:
All good, it was a good call out.  Agreed, I'll make that change.

Reply:
Added an exponential backoff in a recent commit.